File: 240730/수들 중 최솟값 최대화하기/maximin-of-numbers.py
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]

visited = [0] * n
ans = []

def back(sm, total):
    if sm == n:
        ans.append(min(total))
        return

    for i in range(n):
        if visited[i] == 0:
            visited[i] = 1
            back(sm + 1, total + [arr[sm][i]])
            visited[i] = 0

back(0, [])

print(max(ans))

File: trail4/수들 중 최솟값 최대화하기/maximin-of-numbers.java
import java.util.*;
import java.io.*;

public class Main {
    public static int n, ans;
    public static int[] visited;
    public static int[][] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        ans = 0;
        visited = new int[n];
        for (int i = 0; i < n; i++)
        {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++)
            {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        choose(0, Integer.MAX_VALUE);
        System.out.print(ans);
        
    }

    public static void choose(int idx, int minValue)
    {
        if (idx >= n)
        {
            ans = Math.max(ans, minValue);
            return;
        }

        for (int i = 0; i < n; i++)
        {
            if (visited[i] == 1)
            {
                continue;
            }
            visited[i] = 1;
            choose(idx+1, Math.min(minValue, arr[idx][i]));
            visited[i] = 0;
        }
    }
}